internal/component/prometheus/scrape/scrape.go

```diff
@@ -341,6 +352,32 @@ func (c *Component) distributeTargets(
 	// by the scrape loop itself during the sync.
 	promMovedTargets := c.populatePromLabels(movedTargets, jobName, args)
 
+	if c.debugDataPublisher.IsActive(componentID) {
+		var (
+			oldTargetLabels   labels.Labels
+			newTargetLabels   labels.Labels
+			movedTargetLabels labels.Labels
+		)
+		for _, t := range oldDistributedTargets.LocalTargets() {
+			oldTargetLabels = append(oldTargetLabels, t.Labels().Copy()...)
+		}
+
+		for _, t := range newLocalTargets {
+			newTargetLabels = append(newTargetLabels, t.Labels().Copy()...)
+		}
+
+		for _, t := range movedTargets {
+			movedTargetLabels = append(movedTargetLabels, t.Labels().Copy()...)
+		}
+
+		data := fmt.Sprintf("oldTargetLabels: %s => newTargetLabels: %s => movedTargetLabels: %s",
+			oldTargetLabels.String(),
+			newTargetLabels.String(),
+			movedTargetLabels.String(),
+		)
+		c.debugDataPublisher.Publish(componentID, data)
+	}
```

Comment:
with the scrape component it would be more interesting to get the metrics than the targets. I solved this on an experimental branch by using a Prometheus interceptor: https://github.com/grafana/alloy/blob/hackathon-alloy-live-graph/internal/component/prometheus/scrape/scrape.go#L266
I don't know if that's the best approach or not, it's crucial that it does not impact the performances when the debugDataPublisher is not active

Reply:
The interceptor sounds good. It acts a proxy function call and without debug enabled the publish is not going to happen. Also I see that the idea of interceptor is used in remote_write as well.

The other way is to hook up the interceptor only if the debugging is Active but it would make the code complicated as the `Update` function only updates the config and we have scraper manager which runs independently of update.

I feel the interceptor sounds good. Do you have other thoughts ?